flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,458 @@
+- Making it easier to create new capabilities (with individual links).
+- Offering a way to iterate through capabilities on a storage path.
+- Removing the need to have a `/private` domain.
+- Introducing Capability Controllers (CapCons) that handle management of capabilities.
+
+### Capability Controllers (CapCons)
+
+Each Capability would have an associated CapCon that would be responsible for managing the Capability. The CapCon would be created when the Capability is issued (created). If the Capability is copied (since it is a value type) it shares the CapCon with the original (i.e.: calling revoke on the CapCon revokes all copies of the associated capability).
+
+The data associated with CapCons would be stored in arrays, so that each storage path on an account has an array of CapCons of Capabilities issued from that storage path. CapCons are a non-storable object (similar to AuthAccounts).
+
+The definition of the `CapabilityController`.
+
+```cadence
+struct CapabilityController {
+   // The block height when the capability was created.
+   let issueHeight: UInt64
+   // The Type of the capability, i.e.: the T in Capability<T>.
+   let borrowType: Type
+   // The id of the related capability.
+   // This is the UUID of the created capability.
+   // All copies of the same capability will have the same UUID
+   let capabilityID: UInt64
+   // Is the capability revoked.
+   let isRevoked: Bool
+
+   // Returns the targeted storage path of the capability.
+   fun target(): StoragePath
+   // Revoke the capability making it no longer usable.
+   // When borrowing from a revoked capability the borrow returns nil.
+   fun revoke()
+   // Retarget the capability.
+   // This moves the CapCon from one CapCon array to another.
+   // If the new target is not a valid target this will panic.
+   fun retarget(target: StoragePath)
+}
+```
+
+The capability related methods in the `AuthAccount` and in the `PublicAccount` would be moved to a `capabilities` namespace, similar to how the contract methods are in the [contracts namespace](https://developers.flow.com/cadence/language/accounts).
+
+The `AuthAccount` would get new methods to create or get or iterate through capabilities and to get or iterate through CapCons in order to manage capabilities. Methods used for (un)linking would be removed as they are no longer needed.
+
+```cadence
+struct AuthAccount {
+   // ...
+   // removed: fun link<T: &Any>(_ newCapabilityPath: CapabilityPath, target: Path): Capability<T>?
+   // removed: fun getLinkTarget(_ path: CapabilityPath): Path?
+   // removed: fun unlink(_ path: CapabilityPath)
+   // moved & renamed: fun getCapability<T>(_ path: CapabilityPath): Capability<T>
+   // moved & renamed: fun forEachPublic(_ function: ((PublicPath, Type): Bool))
+   let capabilities: AuthAccount.Capabilities
+   struct Capabilities {
+      // get returns the capability at the public path, if one was stored there.
+      fun get<T>(_ path: PublicPath): Capability<T>?
```

Comment:
Another thing I noticed while going over the existing API and the new one:
`link` has a type parameter `<T: &Any>`, i.e. the type parameter has a type bound `&Any`, which basically expresses that type arguments for must be a reference type.
We should maybe have the same requirements for `get`, `borrow`, and `issue`.

While at it, we could maybe also add that type bound to `Capability<T>`?